resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSchedulingRequestAllocator.scala

```diff
@@ -0,0 +1,381 @@
+          // more containers are available than needed for locality, fill in requests for any host
+          for (i <- 0 until (availableContainers - newLocalityRequests.size)) {
+            newLocalityRequests += createSchedulingRequest(resource, rpId = rpId)
+          }
+        } else {
+          val numToCancel = newLocalityRequests.size - availableContainers
+          // cancel some requests without locality preferences to schedule more local containers
+          anyHostRequests.slice(0, numToCancel).foreach { nonLocal =>
+            removeFromOutstandingSchedulingRequests(nonLocal)
+          }
+          if (numToCancel > 0) {
+            logInfo(s"Canceled $numToCancel unlocalized container requests to " +
+              s"resubmit with locality")
+          }
+        }
+        amClient.addSchedulingRequests(newLocalityRequests.asJava)
+      } else if (numPendingAllocate > 0 && missing < 0) {
+        val numToCancel = math.min(numPendingAllocate, -missing)
+        logInfo(s"Canceling requests for $numToCancel executor container(s) to have a new " +
+          s"desired total ${getOrUpdateTargetNumExecutorsForRPId(rpId)} executors.")
+        // cancel pending allocate requests by taking locality preference into account
+        val cancelRequests = (staleRequests ++ anyHostRequests ++ localRequests).take(numToCancel)
+        cancelRequests.foreach(removeFromOutstandingSchedulingRequests)
+      }
+    }
+  }
+
+  def createSchedulingRequest(
+      resource: Resource,
+      nodes: Array[String] = Array.empty,
+      racks: Array[String] = Array.empty,
+      rpId: Int): SchedulingRequest = {
+    val allocationRequestId = nextAllocationRequestId.getAndIncrement()
+    val nodesLocality = if (nodes.nonEmpty) {
+      allocationRequestIdToNodes.put(allocationRequestId, nodes)
+      val nodesConstraints = nodes.map { node =>
+        targetNodeAttribute(NODE, EQ, nodeAttribute("host", node))
+      }
+      if (nodesConstraints.size > 1) {
+        Some(or(nodesConstraints: _*))
+      } else {
+        Some(nodesConstraints.head)
+      }
+    } else {
+      None
+    }
+
+    val racksLocality = if (racks.nonEmpty) {
+      allocationRequestIdToRacks.put(allocationRequestId, racks)
+      val rackConstraints = racks.map { rack =>
+        targetNodeAttribute(NODE, EQ, nodeAttribute("rack", rack))
+      }
+      if (rackConstraints.size > 1) {
+        Some(or(rackConstraints: _*))
+      } else {
+        Some(rackConstraints.head)
+      }
+    } else {
+      None
+    }
+
+    val locality = (nodesLocality, racksLocality) match {
+      case (Some(nodes), Some(racks)) => Some(or(nodes, racks))
+      case (Some(nodes), None) => Some(nodes)
+      case (None, Some(racks)) => Some(racks)
+      case _ => None
+    }
+
+    val expression = (nodeAttributes.isDefined, locality.isDefined) match {
+      case (true, true) => Some(or(and(nodeAttributes.get, locality.get), nodeAttributes.get))
```

Comment:
Here should use  below constraint is better.
```
delayedOr(
timedClockConstraint(and(nodeAttributes, locality), delayedOrIntervalMilliseconds, TimeUnit.MILLISECONDS),
timedClockConstraint(nodeAttributes, delayedOrIntervalMilliseconds * 2, TimeUnit.MILLISECONDS)))
```

But it always stuck when test in yarn-3.3.0 cluster. Hope for help at this part. It's hard to balance node attribute constraint and locality requirement.